sdk/nodejs/runtime/callbacks.ts

```diff
@@ -0,0 +1,409 @@
+                custom: request.getCustom(),
+                type: request.getType(),
+                name: request.getName(),
+                props: deserializePropertyMap(request.getProperties()) || {},
+                opts: ropts,
+            };
+
+            const result = await transform(args);
+
+            const response = new resproto.TransformResponse();
+            if (result === undefined) {
+                response.setProperties(request.getProperties());
+                response.setOptions(request.getOptions());
+            } else {
+                const mprops = await serializeProperties("props", result.props);
+                response.setProperties(gstruct.Struct.fromJavaScript(mprops));
+
+                // Copy the options over.
+                if (result.opts !== undefined) {
+                    opts = new resproto.TransformResourceOptions();
+
+                    if (result.opts.aliases !== undefined) {
+                        const aliases = [];
+                        for (const alias of result.opts.aliases) {
+                            const resolved = await output(alias).promise(true);
+                            if (isUnknown(resolved)) {
+                                // Can't do anything with unknowns on options.
+                                continue;
+                            }
+
+                            if (typeof resolved === "string") {
+                                const a = new aliasproto.Alias();
+                                a.setUrn(resolved);
+                                aliases.push(a);
+                            } else {
+                                const spec = new aliasproto.Alias.Spec();
+                                if (resolved.name !== undefined) {
+                                    spec.setName(resolved.name);
+                                }
+                                if (resolved.type !== undefined) {
+                                    spec.setType(resolved.type);
+                                }
+                                if (resolved.project !== undefined) {
+                                    spec.setProject(resolved.project);
+                                }
+                                if (resolved.stack !== undefined) {
+                                    spec.setStack(resolved.stack);
+                                }
+                                if (resolved.parent !== undefined) {
+                                    if (Resource.isInstance(resolved.parent)) {
+                                        spec.setParenturn(await resolved.parent.urn.promise());
+                                    } else {
+                                        spec.setParenturn(resolved.parent);
+                                    }
+                                }
+                                const a = new aliasproto.Alias();
+                                a.setSpec(spec);
```

Comment:
I do wonder if we should have a shared helper for this, e.g.

https://github.com/pulumi/pulumi/blob/0f8e922c715cbe338f6b182539b03aa289016a32/sdk/nodejs/runtime/resource.ts#L370-L405

Reply:
That's what I ended up doing for dotnet, can do the same here.